File: src/main/java/com/paragon/client/managers/CommandManager.java
package com.paragon.client.managers;

import com.paragon.Paragon;
import com.paragon.api.util.Wrapper;
import com.paragon.client.systems.command.Command;
import com.paragon.client.systems.command.impl.*;
import com.paragon.client.systems.module.impl.misc.Cryptic;
import net.minecraft.util.text.TextComponentString;
import net.minecraft.util.text.TextFormatting;
import net.minecraftforge.client.event.ClientChatEvent;
import net.minecraftforge.common.MinecraftForge;
import net.minecraftforge.fml.common.eventhandler.SubscribeEvent;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.Locale;

/**
 * @author Wolfsurge
 */
public class CommandManager implements Wrapper {

    private String prefix = "$";
    private static String lastCommand = "";

    private final List<Command> commands;

    private final List<String> commonPrefixes = Arrays.asList("/", ".", "*", ";", ",");

    public CommandManager() {
        MinecraftForge.EVENT_BUS.register(this);

        commands = Arrays.asList(
                new ConfigCommand(),
                new HelpCommand(),
                OpenFolderCommand.INSTANCE,
                new SocialCommand(),
                new SyntaxCommand()
        );

        Paragon.INSTANCE.getLogger().info("Loaded Command Manager");
    }

    public void handleCommands(String message, boolean fromConsole) {
        if (message.split(" ").length > 0) {
            boolean commandFound = false;
            String commandName = message.split(" ")[0];

            for (Command command : commands) {
                if (command.getName().equalsIgnoreCase(commandName)) {
                    command.whenCalled(Arrays.copyOfRange(message.split(" "), 1, message.split(" ").length), fromConsole);
                    lastCommand = getPrefix() + message;
                    commandFound = true;
                    break;
                }
            }

            if (!commandFound) {
                sendClientMessage(TextFormatting.RED + "Command not found!", fromConsole);
            }
        }
    }

    public void sendClientMessage(String message, boolean fromConsole) {
        // Only send chat message if the message wasn't sent from the console
        if (!fromConsole) {
            mc.player.sendMessage(new TextComponentString(TextFormatting.LIGHT_PURPLE + "Paragon " + TextFormatting.WHITE + "> " + message));
        }

        Paragon.INSTANCE.getConsole().addLine(TextFormatting.LIGHT_PURPLE + "Paragon " + TextFormatting.WHITE + "> " + message);
    }

    @SubscribeEvent
    public void onChatMessage(ClientChatEvent event) {
        // Check if the message starts with the prefix
        if (event.getMessage().startsWith(getPrefix())) {
            event.setCanceled(true);

            handleCommands(event.getMessage().substring(prefix.length()), false);
        }
    }

    public boolean startsWithPrefix(String message) {
        return message.split(" ")[0].toLowerCase().startsWith(getPrefix().toLowerCase(Locale.ROOT)) || commonPrefixes.contains(message.split(" ")[0].toLowerCase()) || message.startsWith("crypt") && Cryptic.INSTANCE.isEnabled();
    }

    public String getPrefix() {
        return prefix;
    }

    public List<String> getCommonPrefixes() {
        return commonPrefixes;
    }

    /**
     * Gets the commands
     *
     * @return The commands
     */
    public List<Command> getCommands() {
        return commands;
    }

    /**
     * Gets the last command sent
     *
     * @return The last command sent
     */
    public String getLastCommand() {
        return lastCommand;
    }

}

File: src/main/kotlin/com/paragon/impl/managers/CommandManager.kt
package com.paragon.impl.managers

import com.paragon.Paragon
import com.paragon.impl.command.impl.*
import com.paragon.impl.module.misc.Cryptic
import com.paragon.util.Wrapper
import net.minecraft.util.text.TextComponentString
import net.minecraft.util.text.TextFormatting.*
import net.minecraftforge.client.event.ClientChatEvent
import net.minecraftforge.common.MinecraftForge
import net.minecraftforge.fml.common.eventhandler.SubscribeEvent
import java.util.*

/**
 * @author Surge
 */
class CommandManager : Wrapper {

    private val prefix = "$"
    var lastCommand = ""

    val commands = arrayListOf(
        ConfigCommand, CopySkinCommand, HelpCommand, OpenFolderCommand, SaveMapCommand, SocialCommand, SyntaxCommand, SizeCommand, NearestStronghold
    )

    val commonPrefixes = listOf("/", ".", "*", ";", ",") as MutableList<String>

    init {
        MinecraftForge.EVENT_BUS.register(this)
        Paragon.INSTANCE.logger.info("Loaded Command Manager")
    }

    fun handleCommands(message: String, fromConsole: Boolean) {
        if (message.split(" ".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray().isNotEmpty()) {
            var commandFound = false
            val commandName = message.split(" ".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()[0]
            for (command in commands) {
                if (command.name.equals(commandName, ignoreCase = true)) {
                    command.whenCalled(message.split(" ".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray().copyOfRange(1, message.split(" ".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray().size), fromConsole)
                    lastCommand = prefix + message
                    commandFound = true
                    break
                }
            }
            if (!commandFound) {
                sendClientMessage(RED.toString() + "Command not found!", fromConsole)
            }
        }
    }

    fun sendClientMessage(message: String, fromConsole: Boolean) {
        // Only send chat message if the message wasn't sent from the console
        if (!fromConsole) {
            minecraft.player.sendMessage(TextComponentString(LIGHT_PURPLE.toString() + "Paragon " + WHITE + "> " + message))
        }

        Paragon.INSTANCE.console.addLine(LIGHT_PURPLE.toString() + "Paragon " + WHITE + "> " + message)
    }

    @SubscribeEvent
    fun onChatMessage(event: ClientChatEvent) {
        // Check if the message starts with the prefix
        if (event.message.startsWith(prefix)) {
            event.isCanceled = true
            handleCommands(event.message.substring(prefix.length), false)
        }
    }

    fun startsWithPrefix(message: String) = message.split(" ".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()[0].lowercase(Locale.getDefault()).startsWith(prefix.lowercase()) || commonPrefixes.contains(
        message.split(" ".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()[0].lowercase(Locale.getDefault())
    ) || message.startsWith("crypt") && Cryptic.isEnabled // WTF

}